pkg/policy/cell/policy_importer.go

```diff
@@ -0,0 +1,423 @@
+	oldPrefixes := i.updatePrefixes(ctx, updates)
+
+	// Apply changes to the repository.
+	//
+	// As we commit to the policy repository, we must also determine the set of identities
+	// to regenerate here. Identities selected by either outgoing or incoming rules
+	// will have to be regenerated.
+	idsToRegen := &set.Set[identity.NumericIdentity]{}
+	startRevision := i.repo.GetRevision()
+	endRevision := startRevision
+	var oldRuleCnt int
+	for _, upd := range updates {
+		var regen *set.Set[identity.NumericIdentity]
+
+		// The standard case: we have an owning resource, either a k8s object
+		// or a file on disk.
+		if upd.Resource != "" {
+			regen, endRevision, oldRuleCnt = i.repo.ReplaceByResource(upd.Rules, upd.Resource)
+		} else {
+			// otherwise, this is a local API call, and we are replacing by labels.
+			// Compute the set of sets of labels to replace.
+			var replaceLabels []labels.LabelArray
+			if upd.ReplaceByLabels {
+				for _, rule := range upd.Rules {
+					replaceLabels = append(replaceLabels, rule.Labels)
+				}
+			}
+			if len(upd.ReplaceWithLabels) > 0 {
+				replaceLabels = append(replaceLabels, upd.ReplaceWithLabels)
+			}
+
+			if len(upd.Rules) == 0 && len(replaceLabels) == 0 {
+				// No rules, no resource, no labels. This means we should clear all policies.
+				// Add an empty label selector
+				i.log.Info("Policy replace request with no labels, deleting all policies!")
+				replaceLabels = append(replaceLabels, labels.LabelArray{})
+			}
+
+			if len(replaceLabels) >= 0 {
+				i.log.Info("Replacing policy by labels", logfields.Labels, replaceLabels, logfields.Count, len(upd.Rules))
+			}
+			regen, endRevision, oldRuleCnt = i.repo.ReplaceByLabels(upd.Rules, replaceLabels)
+		}
+
+		if len(upd.Rules) == 0 {
+			i.log.Info("Deleted policy from repository",
+				logfields.Resource, upd.Resource,
+				logfields.PolicyRevision, endRevision,
+				"deletedRules", oldRuleCnt,
+				logfields.Identity, slices.Collect(truncate(regen.Members(), 100)))
+		} else {
+			i.log.Info("Upserted policy to repository",
+				logfields.Resource, upd.Resource,
+				logfields.PolicyRevision, endRevision,
+				"deletedRules", oldRuleCnt,
+				logfields.Identity, slices.Collect(truncate(regen.Members(), 100)))
+
+		}
+
+		idsToRegen.Merge(*regen)
+
+		// Report that the policy has been inserted in to the repository.
+		// Paranoia: only do so if this will never block.
+		if upd.DoneChan != nil {
+			select {
+			case upd.DoneChan <- endRevision:
+			default:
+				// should never happen
+				i.log.Warn("BUG: failed to deliver policy apply update notification")
+			}
+		}
+
+		// Send a policy update notification
+		if i.monitorAgent != nil {
+			var msg monitorapi.AgentNotifyMessage
+			if upd.Resource != "" {
+				// The notification can only be tagged with labels, but we don't have them.
+				// Synthesize one for the resource instead.
+				lbls := []string{
+					"cilium.io/resource:" + string(upd.Resource),
+				}
```

Comment:
We can do this for upserts, but for deleting by resource we really need to synthesize some kind of placeholder.

Reply:
OK, this should be fixed.

Reply:
Ah, I didn't realize the labels weren't available during deletion. Under those circumstances, I believe it might be better if the upsert and delete event have the same label for the same policy.

However, it also remains a question if these policy events are needed anymore. They are technically exposed via Hubble, but I'm not aware of any users. Might be simpler to just get rid of them?

Reply:
The correcterest thing to do would be to add a Resource key to the policy event, but those are protobuf definitions and I'm not sure what the commitments are.

v1.16 broke the Delete event anyways, and nobody seems to have noticed :see_no_evil:.

Reply:
Yeah, I think we should remove those agent events (not just the policy ones) eventually. They don't seem to have any consumers but are causing unnecessary computations to be performed.